src/autoscaler/api/server/api_handler_test.go

```diff
@@ -0,0 +1,370 @@
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+
+	"code.cloudfoundry.org/lager"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("ApiHandler", func() {
+	var (
+		bindingdb *fakes.FakeBindingDB
+		handler   *ApiHandler
+		resp      *httptest.ResponseRecorder
+		req       *http.Request
+	)
+	BeforeEach(func() {
+		bindingdb = &fakes.FakeBindingDB{}
+		resp = httptest.NewRecorder()
+
+		handler = NewApiHandler(lager.NewLogger("test"), conf, bindingdb)
+	})
+
+	Describe("GetBrokerCatalog", func() {
+		JustBeforeEach(func() {
+			handler.GetBrokerCatalog(resp, req, map[string]string{})
+		})
+		Context("When getBrokerCatalog is called", func() {
+			It("gets the catalog json", func() {
+				Expect(resp.Code).To(Equal(http.StatusOK))
+				Expect(resp.Body.Bytes()).To(Equal([]byte(catalog)))
+			})
+		})
+	})
+
+	Describe("CreateServiceInstance", func() {
+		JustBeforeEach(func() {
+			handler.CreateServiceInstance(resp, req, map[string]string{"instanceId": "an-instance-id"})
+		})
+		Context("When mandatory parameters are not provided", func() {
+			BeforeEach(func() {
+				instanceCreationReqBody := &models.InstanceCreationRequestBody{
+					OrgGUID: "an-org-guid",
+					BrokerCommonRequestBody: models.BrokerCommonRequestBody{
+						ServiceID: "a-service-id",
+						PlanID:    "a-plan-id",
+					},
+				}
+				body, err := json.Marshal(instanceCreationReqBody)
+				Expect(err).NotTo(HaveOccurred())
+
+				req, err = http.NewRequest(http.MethodPut, "", bytes.NewReader(body))
+			})
+			It("fails with 400", func() {
+				Expect(resp.Code).To(Equal(http.StatusBadRequest))
+			})
+		})
+
+		Context("When all mandatory parameters are present", func() {
+			BeforeEach(func() {
+				instanceCreationReqBody := &models.InstanceCreationRequestBody{
+					OrgGUID:   "an-org-guid",
+					SpaceGUID: "a-space-guid",
+					BrokerCommonRequestBody: models.BrokerCommonRequestBody{
+						ServiceID: "a-service-id",
+						PlanID:    "a-plan-id",
+					},
+				}
+				body, err := json.Marshal(instanceCreationReqBody)
+				Expect(err).NotTo(HaveOccurred())
+
+				req, err = http.NewRequest(http.MethodPut, "", bytes.NewReader(body))
+			})
+			It("succeeds with 201", func() {
+				Expect(resp.Code).To(Equal(http.StatusCreated))
+			})
+		})
+
+		Context("When database CreateServiceInstance call returns ErrAlreadyExists", func() {
+			BeforeEach(func() {
+				instanceCreationReqBody := &models.InstanceCreationRequestBody{
+					OrgGUID:   "an-org-guid",
+					SpaceGUID: "a-space-guid",
+					BrokerCommonRequestBody: models.BrokerCommonRequestBody{
+						ServiceID: "a-service-id",
+						PlanID:    "a-plan-id",
+					},
+				}
+				body, err := json.Marshal(instanceCreationReqBody)
+				Expect(err).NotTo(HaveOccurred())
+
+				req, err = http.NewRequest(http.MethodPut, "", bytes.NewReader(body))
+
+				bindingdb.CreateServiceInstanceReturns(db.ErrAlreadyExists)
+			})
+			It("succeeds with 200", func() {
```

Comment:
The status code should be 409.

Reply:
> 200 SHOULD be returned if the Service Instance already exists, is fully provisioned, and the requested parameters are identical to the existing Service Instance.
https://github.com/openservicebrokerapi/servicebroker/blob/v2.14/spec.md#provisioning